rts/motoko-rts/src/types.rs

```diff
@@ -124,51 +124,136 @@ impl From<Words<u32>> for Bytes<u32> {
     }
 }
 
+/// A value in a heap slot
 #[repr(transparent)]
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub struct SkewedPtr(pub usize);
+pub struct Value(u32);
 
-impl SkewedPtr {
-    pub unsafe fn tag(self) -> Tag {
-        (self.unskew() as *mut Obj).tag()
+/// A view of `Value` for analyzing the slot contents.
+pub enum PtrOrScalar {
+    /// Slot is a pointer to a boxed object
+    Ptr(usize),
+
+    /// Slot is an unboxed scalar value
+    Scalar(u32),
+}
+
+impl PtrOrScalar {
+    pub fn is_ptr(&self) -> bool {
+        matches!(self, PtrOrScalar::Ptr(_))
+    }
+
+    pub fn is_scalar(&self) -> bool {
+        matches!(self, PtrOrScalar::Scalar(_))
+    }
+}
+
+impl Value {
+    /// Create a value from a pointer
+    pub const fn from_ptr(ptr: usize) -> Self {
+        // TODO: Update rustc to enable assertion
+        // debug_assert_eq!(ptr & 0b1, 0b0);
+        Value(skew(ptr) as u32)
+    }
+
+    /// Create a value from a scalar
+    pub const fn from_scalar(value: u32) -> Self {
+        // TODO: Update rustc to enable assertion
+        // debug_assert_eq!(value >> 31, 0);
+        Value(value << 1)
     }
 
-    pub fn unskew(self) -> usize {
-        self.0.wrapping_add(1)
+    /// Create a value from raw representation. Useful when e.g. temporarily writing invalid values
+    /// to object fields in garbage collection.
+    pub const fn from_raw(raw: u32) -> Self {
+        Value(raw)
+    }
+
+    /// Analyzes the value.
+    pub fn get(&self) -> PtrOrScalar {
+        if self.0 & 0b1 == 0b1 {
+            PtrOrScalar::Ptr(unskew(self.0 as usize))
```

Comment:
I assume this is just stack allocation in Rust, right?

Reply:
Yes.